firedrake/slate/preconditioners.py

```diff
@@ -0,0 +1,285 @@
+        sigma_h, u_h = self.broken_solution.split()
+        g, f = self.broken_rhs.split()
+
+        # Pressure reconstruction
+        M = B * A.inv * B.T + C
+        u_sol = M.inv * f + M.inv * (B * A.inv *
+                                     K_local.T * self.trace_solution
+                                     - B * A.inv * g)
+        self._assemble_pressure = create_assembly_callable(
+            u_sol,
+            tensor=u_h,
+            form_compiler_parameters=context.fc_params)
+
+        # Velocity reconstruction
+        sigma_sol = A.inv * g + A.inv * (B.T * u_h -
+                                         K_local.T * self.trace_solution)
+        self._assemble_velocity = create_assembly_callable(
+            sigma_sol,
+            tensor=sigma_h,
+            form_compiler_parameters=context.fc_params)
+
+        # Set up the projector for projecting the broken solution
+        # into the unbroken finite element spaces
+        # NOTE: Tolerance here matters!
+        sigma_b, _ = self.broken_solution.split()
+        sigma_u, _ = self.unbroken_solution.split()
+        self.projector = Projector(sigma_b,
+                                   sigma_u,
+                                   solver_parameters={"ksp_type": "cg",
+                                                      "ksp_rtol": 1e-13})
+
+    def update(self, pc):
+        """Update by assembling into the operator. No need to
+        reconstruct symbolic objects.
+        """
+        self._assemble_S()
+        self.S.force_evaluation()
+        self._assemble_Srhs()
+
+    def apply(self, pc, x, y):
+        """We solve the forward eliminated problem for the
+        approximate traces of the scalar solution (the multipliers)
+        and reconstruct the "broken flux and scalar variable."
+
+        Lastly, we project the broken solutions into the mimetic
+        non-broken finite element space.
+        """
+        from firedrake import project
+
+        # Transfer non-broken x into a firedrake function
+        with self.unbroken_rhs.dat.vec as v:
+            x.copy(v)
+
+        # Transfer unbroken_rhs into broken_rhs
+        field0, field1 = self.unbroken_rhs.split()
+        bfield0, bfield1 = self.broken_rhs.split()
+
+        # This updates broken_rhs
+        project(field0, bfield0)
+        field1.dat.copy(bfield1.dat)
+
+        # Compute the rhs for the multiplier system
+        self._assemble_Srhs()
+
+        # Solve the system for the Lagrange multipliers
+        with self.schur_rhs.dat.vec_ro as b:
+            with self.trace_solution.dat.vec as x:
+                self.ksp.solve(b, x)
+
+        # Assemble the pressure and velocity (in that order)
+        # using the Lagrange multipliers
+        self._assemble_pressure()
+        self._assemble_velocity()
+
+        # Project the broken solution into non-broken spaces
+        sigma_h, u_h = self.broken_solution.split()
+        sigma_u, u_u = self.unbroken_solution.split()
```

Comment:
Inconsistency with naming before.  Also with the naming of the spaces.  Perhaps:

```python
sigma_d, u_d = self.broken_solution.split()
sigma, u = self.unbroken_solution.split()
```

Warning, implicit assumption that you have `V*P`.